### PR TITLE
Live Preview: apply selected style variation to the preview

### PR DESCRIPTION
**Pull request by fushar**
Related to:

- https://github.com/Automattic/wp-calypso/issues/82914

## Proposed Changes

This PR supports applying the selected style variation to the Block Theme Previews. This works by updating the previewed theme's global styles CPT (post with type=`wp_global_styles`) (creating it first if it does not exist already) with the selected style variation. Note that even without this feature, a global styles CPT will be always created when a user live-preview a theme.

## Testing Instructions

1. Start with a site with Free plan. It's best to use a brand new site, so that it only has global styles CPT of the currently active theme.
2. Preview a theme WITHOUT style variations; verify that it still works.
3. Preview a theme with style variations, but without selecting any non-default variation; verify that it still works.
4. Preview the above theme but with a preselected non-default variation; verify that the variation is applied when the Site Editor loads.
5. Verify that the preselected variation is also selected in the `Styles` menu from the left sidebar:
    image
6. Preview the same theme but now with a default variation; verify that it still works (the default variation is applied).
7. Preview the same theme with another different non-default variation. Activate the theme.
8. Activate another theme.
9. Activate the previous theme WITHOUT selecting any variation; verify that you see the previous variation.
10. Upgrade the site to be an Atomic site.
11. Repeat and verify the steps above.

## Pre-merge Checklist

- [x] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [x] https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md for your changes?
- [x] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [x] Have you checked for TypeScript, React or other console errors?
- [x] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [x] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [x] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

### CODE REVIEW

File: client/state/themes/actions/live-preview.ts [hunk 1]
```diff
@@ -22,8 +37,18 @@ export function livePreview( themeId: string, siteId: number, source?: 'list' |
...
+			await dispatch( installTheme( installId, siteId ) );
 		}
+
+		if ( styleVariation ) {
+			const stylesheet = isJetpackSite( getState(), siteId )
+				? themeId
+				: getCanonicalTheme( getState(), siteId, themeId )?.stylesheet ?? '';
+
+			const globalStylesId = await dispatch( upsertAndGetGlobalStylesId( siteId, stylesheet ) );
+			dispatch( updateGlobalStyles( siteId, globalStylesId, styleVariation ) );
```
Comment (arthur791004) [inline client/state/themes/actions/live-preview.ts]: I think we can wait until the request is done before redirecting to the live preview. WDYT?

File: client/state/global-styles/actions.ts [hunk 2]
```diff
@@ -1,23 +1,26 @@
 import wpcom from 'calypso/lib/wp';
 import type { ActiveTheme, GlobalStyles } from '@automattic/data-stores';
 
+export function upsertAndGetGlobalStylesId( siteIdOrSlug: number | string, stylesheet: string ) {
```
Comment (arthur791004) [inline client/state/global-styles/actions.ts]: what does the `upsert` mean here?

File: client/state/global-styles/actions.ts [hunk 3]
```diff
@@ -1,23 +1,26 @@
...
+	return async () => {
+		const themes: ActiveTheme[] = await wpcom.req.get( {
+			path: `/sites/${ encodeURIComponent( siteIdOrSlug ) }/themes?${ new URLSearchParams( {
+				status: 'active',
+				wp_theme_preview: stylesheet,
```
Comment (arthur791004) [inline client/state/global-styles/actions.ts:9]: I think we can keep using the `getGlobalStylesId` function but replace the endpoint with this one. WDYT?

**Comment by arthur791004** [comment]
We have to apply the default style variation for the activation or preview as people may preview a non-default style variation first, and then go back to the Theme Details page to activate/preview a default style variation.

**Comment by fushar** [comment]
> We have to apply the default style variation for the activation or preview as people may preview a non-default style variation first, and then go back to the Theme Details page to activate/preview a default style variation.

@arthur791004, but if we do this, then the user will never be able to retrieve the styles from a previously activated theme, which contradicts our [support doc](https://wordpress.com/support/themes/changing-themes/) which we show in the modal when we activate a theme: 

> What Happens to Your Old Content
> Changing themes is a safe, non-destructive action. Your existing pages, media, and other data remain on the site.

> Your site’s homepage will update to the homepage of the new theme. If you wish to recover any content from your previous theme’s homepage, reactivate the previous theme to view and copy any of that content.

That's why I exclude the default variation from the logic. What do you think?

**Comment by arthur791004** [comment]
> but if we do this, then the user will never be able to retrieve the styles from a previously activated theme, which contradicts our [support doc](https://wordpress.com/support/themes/changing-themes/) which we show in the modal when we activate a theme:

The styles follow the theme (active one or the previewing one), so it won't affect the styles of the previously active theme.

Currently, the issue is
1. Preview a non-default style variation
2. We apply the style variation before going to the live preview
3. When the user comes back to the Theme Details page
4. Preview a default style variation or activate the theme with the default style variation
5. When the user lands on the editor, the styles are still the previous one (the non-default style variation for the preview earlier)

It's weird that the styles don't match the one they selected on the Theme Details page 😅

If you mean the user applied styles to a theme (A) before, and then they switched to another theme (B). When they go to preview a theme (A) with the default styles, we have to show the styles they applied before. However, when you apply the non-default styles to that theme (A), those styles won't be recoverable anymore (you have to go to revisions to restore them).

**Comment by fushar** [comment]
> If you mean the user applied styles to a theme (A) before, and then they switched to another theme (B). When they go to preview a theme (A) with the default styles, we have to show the styles they applied before.

Yes, this case is what exactly I mean! I believe this case is quite common (activating a previous theme). It is weird to revert to the default variation when a user activates their previous theme.

> When the user lands on the editor, the styles are still the previous one (the non-default style variation for the preview earlier)
> It's weird that the styles don't match the one they selected on the Theme Details page 😅

Yeah, looks like we cannot satisfy both the above cases using the same logic... but I don't know how to resolve this 😭

**Comment by arthur791004** [comment]
> Yes, this case is what exactly I mean! I believe this case is quite common (activating a previous theme). It is weird to revert to the default variation when a user activates their previous theme.

Yep. But with this change, if the user has previewed a non-default style variation, we're not able to restore previous styles. As a result, no matter whether you exclude the default style variation from this feature or not, we still overwrite the user's global styles when the user live-preview a previously activated theme in the past 🥲

**Comment by fushar** [comment]
Or maybe we can use this logic only if the previewed theme has not been activated previously? Is it possible to check this condition? 😅

**Comment by taipeicoder** [comment]
@fushar we also need to support this when selecting a style variation from a card in the Theme Showcase, using the calypso.live link I get this error 🥲:

Screenshot 2024-01-05 at 9 16 02 AM

**Comment by taipeicoder** [comment]
> Yes, this case is what exactly I mean! I believe this case is quite common (activating a previous theme). It is weird to revert to the default variation when a user activates their previous theme.

If I understood this discussion correctly, "fixing" the Live Preview for the default style variation would mean that:

1. User used theme A.
2. Theme A had custom styles.
3. User switches the theme B.
4. User Live Previews theme A.
5. User switched back to theme A.
6. Theme A custom styles will be overridden with the default styles instead of preserving its custom styles.

Is that correct?

**Comment by fushar** [comment]
> If I understood this discussion correctly, "fixing" the Live Preview for the default style variation would mean that:

Yes, that's correct as far as this feature is concerned. However, this would mean that reactivating an old theme (that happens to have style variations) will always override the past global styles set by the user in that theme, which is probably not what we want 🥲

**Comment by taipeicoder** [comment]
> However, this would mean that reactivating an old theme (that happens to have style variations) will always override the past global styles set by the user in that theme

Is this also the case when the user doesn't Live Preview the old theme and directly activates it?

**Review by arthur791004 — COMMENTED** [review]
Forgot to submit 😅

**Comment by fushar** [comment]
> Is this also the case when the user doesn't Live Preview the old theme and directly activates it?

No, if I activate an old theme but didn't preselect a variation (using the default one), I get my old global styles, not the "default" one. (I just tried it.)

**Comment by taipeicoder** [comment]
My two cents is that we should have the default style variation override GS, even if it means that it will override it for the old active theme (when Live Previewed). The reasoning is that the latter feels (no data to back me up 😅) less likely to happen, and if it does, then it's in theory recoverable via styles revisions.

Screenshot 2024-01-05 at 10 48 19 AM

@Automattic/lego so that we are all looped in on this.

File: client/state/themes/actions/live-preview.ts (hunk 1 above)
Comment (fushar) [reply client/state/themes/actions/live-preview.ts]: This is actually intentional, as it takes quite a long to update global styles (1.5 - 2 seconds), and the redirection always takes more than that (in my testing so far). So we actually get rid of the endpoint call overhead for free by not `await`-ing for the request.

What do you think?

**Comment by fushar** [comment]
@taipeicoder Just to be sure, are you suggesting the following:

- When I directly activate a (possibly old) theme:
   - with default variation: I get my old GS back
   - with a preselected custom variation: I get the custom variation
- When I live-preview a (possibly old) theme (and possibly activate it afterwards):
   - with default variation: my old GS is cleared (thus I get the default variation)
   - with a preselected custom variation: I get the custom variation

?

**Comment by arthur791004** [comment]
> it's in theory recoverable via styles revisions.

Agreed 👍

**Comment by taipeicoder** [comment]
> @taipeicoder Just to be sure, are you suggesting the following:
> 
> * When I directly activate a (possibly old) theme:
>   
>   * with default variation: I get my old GS back
>   * with a preselected custom variation: I get the custom variation
> * When I live-preview a (possibly old) theme (and possibly activate it afterwards):
>   
>   * with default variation: my old GS is cleared (thus I get the default variation)
>   * with a preselected custom variation: I get the custom variation
> 
> ?

Yes! The only scenario where users will get unexpected results is when they switch back to a previously activated theme with custom styles, and had Live Previewed it with either default or non-default style variations.

I counted both default and non-default style variations because if they Live Preview with a style variation and decided they  liked their old one better (they must have a good memory), the old styles will still be overridden 🙂

**Comment by fushar** [comment]
> @fushar we also need to support this when selecting a style variation from a card in the Theme Showcase, using the calypso.live link I get this error 🥲:

@taipeicoder I cannot reproduce this, can you provide repro steps?

**Comment by fushar** [comment]
> > @fushar we also need to support this when selecting a style variation from a card in the Theme Showcase, using the calypso.live link I get this error 🥲:
> 
> @taipeicoder I cannot reproduce this, can you provide repro steps?

Sorry, I see now. The three-dot menu not available by default in local dev, will fix this 😭

File: client/state/global-styles/actions.ts (hunk 3 above)
Comment (fushar) [reply client/state/global-styles/actions.ts:9]: Makes sense, done!

**Comment by fushar** [comment]
> @fushar we also need to support this when selecting a style variation from a card in the Theme Showcase

@taipeicoder I tried but it is really complicated to pass things around in `theme-options.js` 😭 😭

As the three-dot menu is going away very soon, I think it is not worth it to pursue this. I "fixed" this instead via https://github.com/Automattic/wp-calypso/pull/85945/commits/4052160a4a7acb28182f88fdd2bc62c0cadd2549 so that at least it does live-preview without any variation. Let me know what you think 😅

**Comment by fushar** [comment]
@Automattic/lego Thanks for the healthy discussions! I've updated the PR and testing instructions 🙏

**Comment by miksansegundo** [comment]
❌ Preview the above theme but with a preselected non-default variation; verify that the variation is applied when the Site Editor loads.

Using the Calypso Live link, I created a new site, preselected a non-default variation, and clicked "Preview & Customize", but the style variation was not applied. What am I missing?


https://github.com/Automattic/wp-calypso/assets/1881481/9108bc53-1a48-4cac-bfad-8e6a58698602

**Comment by fushar** [comment]
> Using the Calypso Live link, I created a new site, preselected a non-default variation, and clicked "Preview & Customize", but the style variation was not applied. What am I missing?

@miksansegundo See this comment: https://github.com/Automattic/wp-calypso/pull/85945#issuecomment-1878403600

Could you test from theme details for the time being? I can fix the logic on three-dot menu (in a follow-up PR) if we strongly want to fix it 😭

**Comment by taipeicoder** [comment]
> > @fushar we also need to support this when selecting a style variation from a card in the Theme Showcase
> 
> @taipeicoder I tried but it is really complicated to pass things around in `theme-options.js` 😭 😭
> 
> As the three-dot menu is going away very soon, I think it is not worth it to pursue this. I "fixed" this instead via [4052160](https://github.com/Automattic/wp-calypso/commit/4052160a4a7acb28182f88fdd2bc62c0cadd2549) so that at least it does live-preview without any variation. Let me know what you think 😅

@fushar yes let's not adding this behavior to the theme cards for now. We can revisit this decision in case the three-dot menu ends up staying, but let's hope it goes away for good 🙂

File: client/state/themes/actions/live-preview.ts [hunk 4]
```diff
@@ -1,15 +1,28 @@
...
-	return ( dispatch: CalypsoDispatch, getState: () => AppState ) => {
+export function livePreview(
+	siteId: number,
+	themeId: string,
+	hasStyleVariations: boolean,
+	styleVariation?: GlobalStyles,
+	source?: 'list' | 'detail'
+) {
+	return async ( dispatch: CalypsoDispatch, getState: () => AppState ) => {
 		const analysis = recordTracksEvent( 'calypso_block_theme_live_preview_click', {
```
Comment (arthur791004) [inline client/state/themes/actions/live-preview.ts:26]: It would be better to track the style variation 🙂

**Review by arthur791004 — APPROVED** [review]
Tested and looks good to me 🎉

**Comment by arthur791004** [comment]
Proposing https://github.com/Automattic/wp-calypso/pull/86158 to apply the selected style variations from the three-dot menu 🙂

File: client/state/themes/actions/live-preview.ts (hunk 1 above)
Comment (arthur791004) [reply client/state/themes/actions/live-preview.ts]: A concern is people may land on the Site Editor before the global styles are updated successfully.

File: client/state/themes/actions/live-preview.ts (hunk 4 above)
Comment (fushar) [reply client/state/themes/actions/live-preview.ts:26]: Ah, thanks for catching this. I'll add that (maybe in a `theme_style` prop) in a follow-up PR!

File: client/state/themes/actions/live-preview.ts (hunk 1 above)
Comment (fushar) [reply client/state/themes/actions/live-preview.ts]: I've tried this hundreds (kidding!) of times and haven't encountered this issue. Can we revisit this when this actually happen?

The reason why I keep pushing for parallel update is that the delay is pretty bad: around 1.5 - 2.5 seconds (at least in APAC). When a user wants to live-preview multiple themes ("theme shopping") this delay is very annoying and they can perceive WP.com as a slow product...

**Comment by fushar** [comment]
I'm going to ship this now. Wish me luck!

**Comment by autumnfjeld** [comment]
> I'm going to ship this now. Wish me luck!

LOL.  How did the luck work out for you?  🤣

**Pull request merged**